src/Infrastructure.VS.UnitTests/VsUIServiceOperationTests.cs

```diff
@@ -0,0 +1,319 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using FluentAssertions;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Moq;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.TestInfrastructure;
+using Task = System.Threading.Tasks.Task;
+
+namespace SonarLint.VisualStudio.Infrastructure.VS.UnitTests
+{
+    [TestClass]
+    public class VsUIServiceOperationTests
+    {
+        // Dummy interfaces used in the tests
+#pragma warning disable IDE1006 // Naming Styles
+        public interface SMyInterface { }
+#pragma warning restore IDE1006 // Naming Styles
+        public interface IMyInterface { }
+
+        #region MEF tests
+
+        [TestMethod]
+        public void MefCtor_CheckIsExported()
+            => MefTestHelpers.CheckTypeCanBeImported<VsUIServiceOperation, IVsUIServiceOperation>(
+                MefTestHelpers.CreateExport<SVsServiceProvider>(),
+                MefTestHelpers.CreateExport<IThreadHandling>());
+
+        [TestMethod]
+        public void CheckIsNonSharedMefComponent()
+            => MefTestHelpers.CheckIsNonSharedMefComponent<VsUIServiceOperation>();
+
+        [TestMethod]
+        public void MefCtor_DoesNotCallAnyServices()
+        {
+            var serviceProvider = new Mock<IServiceProvider>();
+            var threadHandling = new Mock<IThreadHandling>();
+
+            _ = CreateTestSubject(serviceProvider.Object, threadHandling.Object);
+
+            // The MEF constructor should be free-threaded, which it will be if
+            // it doesn't make any external calls.
+            serviceProvider.Invocations.Should().BeEmpty();
+            threadHandling.Invocations.Should().BeEmpty();
+        }
+
+        #endregion
+
+        #region RunOnUIThread tests
+
+        [TestMethod]
+        public void Execute_Action_RunOnUIThread()
+        {
+            (var calls, var testSubject) = SetupUpThreadHandlingTests<SMyInterface>();
+
+            var callback = (IMyInterface x) => calls.Add("callback");
+
+            // Act
+            testSubject.Execute<SMyInterface, IMyInterface>(callback);
+
+            calls.Should().ContainInOrder(
+                "change thread - sync",
+                "get service",
+                "callback");
+        }
+
+        [TestMethod]
+        public void Execute_Func_RunOnUIThread()
+        {
+            (var calls, var testSubject) = SetupUpThreadHandlingTests<SMyInterface>();
+
+            string callback(IMyInterface x)
```

Comment:
FYI I had originally defined this as a `Func<IMyInterface>` variable, but an IDE quick fix suggested changing it to a local function instead.